draft-ietf-quic-transport.md

```diff
@@ -3257,6 +3287,49 @@ the Source Connection ID used by the peer during the initial
 handshake, it MUST treat that receipt as a connection error of type
 PROTOCOL_VIOLATION.
 
+## CONNECTION_ID_FINISHED Frame {#frame-connection-id-finished}
+
+An endpoint sends a CONNECTION_ID_FINISHED frame (type=0x1b) to indicate that it
+will no longer use a connection ID that was issued by its peer.  This also
+serves as a request to the peer to send additional connection IDs for future use
+(see {{connection-id}}).  New connection IDs can be delivered via the
+NEW_CONNECTION_ID frame ({{frame-new-connection-id}}).
+
+Retiring a connection ID using the CONNECTION_ID_FINISHED frame invalidates any
+stateless reset tokens associated with that connection ID.
+
+The CONNECTION_ID_FINISHED frame is as follows:
+
+~~~
+ 0                   1                   2                   3
+ 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1
++-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+|   Length (8)  |          Connection ID (32..144)            ...
++-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+~~~
+
+The fields are:
+
+Length:
+
+: An 8-bit unsigned integer containing the length of the connection ID.
+
+Connection ID:
+
+: A connection ID of the specified length.
+
+Receipt of a CONNECTION_ID_FINISHED frame containing a connection ID that was
+not previously sent to the peer MAY be treated as a connection error of type
```

Comment:
Umm aren't you reversing direction of CONNECTION_ID_FINISHED. One end sends ID's, the other end finishes them so ACK shouldn't matter. On the MAY: if the endpoint tracks generated CID's via an algorithm such as hashing a counter. With relatively short CIDs there are possible collisions and you might not want to actually store the CIDs already generated.

Reply:
Yeah, @mikkelfj got it.  You can retire a connection ID before NEW_CONNECTION_ID is acknowledged even.

As for the short connection IDs, I think that's a reason not to allow shorter values more than an argument against this text.  Yes, over time, you might cycle really old IDs back into circulation such that they get re-issued to a connection that had it in the past.  That would be valid, but allowing it would complicate the protocol.  The protocol here uses the connection ID as its own identifier, so for that to work properly, we need it to be unique.